Mathlib/LinearAlgebra/Matrix/DFT.lean

```diff
@@ -0,0 +1,274 @@
+/-
+Copyright (c) 2020 Mohanad Ahmed. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mohanad Ahmed
+-/
+
+import Mathlib.Algebra.GeomSum
+import Mathlib.Data.Complex.Basic
+import Mathlib.Data.Complex.Exponential
+import Mathlib.Analysis.SpecialFunctions.Complex.Log
+import Mathlib.Analysis.SpecialFunctions.Trigonometric.Basic
+import Mathlib.LinearAlgebra.Matrix.NonsingularInverse
+import Mathlib.LinearAlgebra.Matrix.Circulant
+import Mathlib.LinearAlgebra.Vandermonde
+
+/-!
+# Discrete Fourier Transform (DFT) Matrix and DFT of a (finite) sequence
+
+This file defines the `dft` opertaion on a sequence (also a vecotr) and the DFT operation matrix
+
+## Main definitions
+
+ - `dft v`: given a sequence (v : (Fin n) → ℂ) we can transform it into a sequence (V : (Fin n) →ℂ)
+ such that
+ $$ V [p] = ∑_{k = 0}^{N - 1} e^{-j*2πkp/n} v [k] $$
+ - `idft V` : given a sequence (V : (Fin n) → ℂ) we can transform it into a sequence
+ (v : (Fin n) → ℂ)
+such that
+$$ v [k] = ¼{1}{N}∑_{p = 0}^{n - 1} e^{j*2πkp/N} v [p] $$
+- `Wₙ` : the dft matrix with the `k, p` entry equal to
+$$ Wₙ[k, p] = e^{-j2πkp/n} $$
+- `Wₙ⁻¹` : the idft matrix with the `k, p` entry equal to
+$$ Wₙ[k, p] = (1/N) * e^{j2πkp/n} $$
+
+
+## Main results
+
+- `dft v = Wₙ v` : the dft operation on a sequence is the same as the dft matrix applied to the
+vector
+- `idft V = Wₙ⁻¹ V` : the idft operation on a sequence is the same as the idft matrix applied to the
+vector
+- `dft (idft v) = dft ( idft v) = v` the dft and idft operations are inverses
+- `Wₙ = vandermonde (w)` : the dft matrix is vandermonde with `w` being the first row of the dft
+matrix
+- `circulant t = Wₙ⁻¹ ⬝ diagonal (dft t) ⬝ Wₙ` : a circulant matrix is diagonalizable by the dft and
+idft matrix pair.
+
+-/
+
+namespace DFT
+
+
+open Complex Matrix BigOperators Finset Real
+
+variable (n : ℕ) [NeZero n]
```

Comment:
I don't see why you require `NeZero`, I know mathematically it doesn't matter but it will make things more annoying to use at the point of use